Instrument mock operations

heaths opened the pull request:
Resolves #27393

Previously non-instrumented operations were potentially regressed in PR #27368 and would instead run with their default polling of 1s if not specified by a header. Since Key Vault uses the standard `Retry-After` header with a 1s resolution, one viable alternative was to instrument operations from `ClientTestBase` similar to clients themselves. The code is effectively the same as in `RecordedTestBase` but without the dependency of using recorded tests.

heaths committed:
Instrument mock operations

Resolves #27393

m-nash commented on sdk/keyvault/Azure.Security.KeyVault.Certificates/tests/CertificateOperationTests.cs:
```diff
@@ -115,6 +116,7 @@ public async Task UpdateStatusEventuallyCompleted()
 
             CertificateClient client = CreateClient(transport);
             CertificateOperation operation = await client.StartCreateCertificateAsync(CertificateName, s_policy);
+            operation = InstrumentMockOperation(operation);
```
I think we should investigate why the operation is not automatically instrumented here.  For instance any method returning an [Operation like this](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/tests/Scenario/ResourceGroupCollectionTests.cs#L150) gets automatically instrumented by the test framework.  I believe the code that does that[ is here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/Instrumentation/InstrumentResultInterceptor.cs#L61) but for some reason it isn't catching this case and we should probably update it so that we don't have to manually find all of these.

heaths replied:
Because my test class only extends `ClientTestBase` and operations only get instrumented by `RecordedTestBase`:

https://github.com/Azure/azure-sdk-for-net/blob/b732adadb71dadfee5c083165b90e1d6a3148ffb/sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs#L352-L353

The `ClientTestBase.InstrumentOperation` merely returns the operation unchanged. As noted in my [reply below](https://github.com/Azure/azure-sdk-for-net/pull/27399#discussion_r821914946) I'm not keen on regressing any of the many track 2 libraries might be inadvertently taking advantage of this fact, so I made a new method anyone needing to mock operations without using the recording infrastructure - which is a non-starter for this test: I can't control what the service does but want to test a particular chain of responses.

/cc @JoshLove-msft

heaths replied:
According to these [search results](https://cs.github.com/Azure/azure-sdk-for-net?q=InstrumentOperation+ClientTestBase), outside of these related base classes, I don't see any. Everything else seems to extend from `RecordedTestBase`, including some even more-derived classes in Azure.ResourceManager.Network.

m-nash commented on sdk/core/Azure.Core.TestFramework/src/ClientTestBase.cs:
```diff
@@ -143,6 +144,22 @@ protected internal virtual object InstrumentOperation(Type operationType, object
             return operation;
         }
 
+        protected internal T InstrumentOperation<T>(T operation) where T : Operation =>
```
I believe these methods were left off of here because we shouldn't be adding the OperationInstrumentor when we are not in playback mode and playback mode was assumed to be only for recorded tests.  We might want to come up a MockTestBase that allows us to maintain the enforcement of only applying the operation instrumentation during playback but at the same time allows us to override that only in the Mock scenario.  It would probably also be worth it to add a mock instrumentor to enforce that if your test base doesn't inherit from the mocktestbase then your transport shouldn't be a mocktransport and vice versa.

heaths replied:
These methods aren't called by default so I see no harm is refactoring them from `RecordedTestBase` to `ClientTestBase`. Could you clarify why you are concerned where they are defined when they have to be manually called?

heaths replied:
See https://github.com/Azure/azure-sdk-for-net/pull/27399#discussion_r821914946

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/ClientTestBase.cs:
```diff
@@ -143,6 +144,22 @@ protected internal virtual object InstrumentOperation(Type operationType, object
         }
 
+        protected internal T InstrumentOperation<T>(T operation) where T : Operation =>
+            (T)InstrumentOperation(typeof(T), operation);
```
Why did we need to move this method up to ClientTestBase?

heaths replied:
Because I use it in a class derived from `ClientTestBase`, not `RecordedTestBase`.

JoshLove-msft commented on sdk/core/Azure.Core.TestFramework/src/ClientTestBase.cs:
```diff
@@ -143,6 +144,22 @@ protected internal virtual object InstrumentOperation(Type operationType, object
+        protected internal T InstrumentOperation<T>(T operation) where T : Operation =>
+            (T)InstrumentOperation(typeof(T), operation);
+
+        protected object InstrumentMockOperation(Type operationType, object operation, params IInterceptor[] interceptors)
```
Why not just use the same name and have RecordedTestBase override this one. That way, we don't have to worry about recorded tests calling this method (to Michael's point about validation).

heaths replied:
I'm not sure to which method you're referring since you have two in context, but to both you and @m-nash:

1. I'm not setting out to regress any functionality with classes that already extend `ClientTestBase` and use `InstrumentOperation` already. That's why I added a new method that was specific to mocking, since after the regressing PR we can't simply use a `MockTransport` et. al. with operations.
2. The generic "overload" was for ease. Not sure why a method that actually takes a `Type` parameter was ever added, but it's more convenience - and safe, given we can specify generic type constraints - to have a generic method. `RecordedTestBase` already had it and would've used a `virtcall` to call the virtual `InstrumentOperation(Type, ...)` method, and still will. Where this method was defined doesn't change the fact it still uses a `virtcall`. The behavior will be unchanged. I simply refactored the method up the hierarchy.

JoshLove-msft replied:
I'm suggesting that we don't add a separate InstrumentMockOperation method but instead add an implentation to the existing InstrumentOperation method. We can check to see if there are any such usages that you mention. The benefit of doing this is that it would make it impossible to call the wrong InstrumentOperation method because recorded tests will always use the one defined in RecordedTestBase.

heaths replied:
According to these [search results](https://cs.github.com/Azure/azure-sdk-for-net?q=InstrumentOperation+ClientTestBase), outside of these related base classes, I don't see any. Everything else seems to extend from `RecordedTestBase`, including some even more-derived classes in Azure.ResourceManager.Network.

Perhaps this could work then, and that should solve the [concern](https://github.com/Azure/azure-sdk-for-net/pull/27399#discussion_r821253497) @m-nash raised.

heaths committed:
Resolve PR feedback

heaths commented on sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs:
```diff
@@ -344,23 +344,11 @@ protected internal override object InstrumentClient(Type clientType, object clie
-                operation,
-                interceptorArray);
-        }
+            => InstrumentOperationInternal(operationType, operation, Mode == RecordedTestMode.Playback, interceptors);
```
The only other code according to [code search](https://cs.github.com/Azure/azure-sdk-for-net?q=InstrumentOperationInternal) that calls this I could also change to just call the `ClientTestBase.InstrumentOperationInternal`, but I felt like this helped codify a `RecordedTestBase`-specific variant with the right condition. I'm not strongly in favor of keeping this or changing the seemingly sole caller either way, though.

/cc @JoshLove-msft @m-nash

JoshLove-msft replied:
I think it is fine to leave like this.

heaths commented:
/azp run net - appconfiguration - tests

heaths committed:
Resolve offline feedback

heaths commented:
/azp run net - appconfiguration - tests

heaths commented:
Blocked on PR #26676

heaths commented:
The test `SkipsUpdateWithNoChangesNull("PocoTableEntity")` timed out but was normally around ~1s, which was on par with other platforms:

<img width="127" alt="image" src="https://user-images.githubusercontent.com/1532486/157978431-5a2d13c7-0ae8-4231-8884-f5786425dc9e.png">

However, it doesn't seem like this test even uses LROs so it may just be a fluke: https://github.com/Azure/azure-sdk-for-net/blob/b1b3e17312e4180b1b913c87a50c5ec6df4ab973/sdk/tables/Microsoft.Azure.WebJobs.Extensions.Tables/tests/EntityBindingLiveTests.cs#L824

Re-running the failed job to see if it repros; otherwise, we'll just have to keep an eye on it.

/cc @christothes

heaths committed:
Ignore unrelated failing tests

Opened #27522 to track issue in Azure.Communication.NetworkTraversal.
Changes were made since the merge base which seem to be related.
